plugins/plant.cpp

```diff
@@ -0,0 +1,627 @@
+    { struct_field_info::PRIMITIVE, "zlevel",    offsetof(plant_options, zlevel),    &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "dry_run",   offsetof(plant_options, dry_run),   &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "plant_idx", offsetof(plant_options, plant_idx), &df::identity_traits<int32_t>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "age",       offsetof(plant_options, age),       &df::identity_traits<int32_t>::identity, 0, 0 },
+    { struct_field_info::END }
+};
+struct_identity plant_options::_identity(sizeof(plant_options), &df::allocator_fn<plant_options>, NULL, "plant_options", NULL, plant_options_fields);
+
+const int32_t sapling_to_tree_threshold = 120 * 28 * 12 * 3 - 1; // 3 years minus 1; let the game handle the actual growing-up
+
+command_result df_createplant(color_ostream &out, df::coord pos, const plant_options &options)
+{
+    auto col = Maps::getBlockColumn((pos.x / 48)*3, (pos.y / 48)*3);
+    if (!Maps::getTileBlock(pos) || !col)
+    {
+        out.printerr("No map block at pos!\n");
+        return CR_FAILURE;
+    }
+
+    auto tt = Maps::getTileType(pos);
+    if (!tt || tileShape(*tt) != tiletype_shape::FLOOR)
+    {
+        out.printerr("Plants can only be placed on floors!\n");
+        return CR_FAILURE;
+    }
+    else // Check tile mat and building occ
+    {
+        auto mat = tileMaterial(*tt);
+        if (mat != tiletype_material::SOIL &&
+            mat != tiletype_material::GRASS_DARK &&
+            mat != tiletype_material::GRASS_LIGHT)
+        {
+            out.printerr("Plants can only be placed on dirt or grass!\n");
+            return CR_FAILURE;
+        }
+
+        auto occ = Maps::getTileOccupancy(pos);
+        if (occ && occ->bits.building > tile_building_occ::None)
+        {
+            out.printerr("Building present at pos!\n");
+            return CR_FAILURE;
+        }
+    }
+
+    auto p_raw = vector_get(world->raws.plants.all, options.plant_idx);
+    if (!p_raw)
+    {
+        out.printerr("Plant raw not found!\n");
+        return CR_FAILURE;
+    }
+
+    auto plant = df::allocate<df::plant>();
+    if (p_raw->flags.is_set(plant_raw_flags::TREE))
+        plant->hitpoints = 400000;
+    else
+    {
+        plant->hitpoints = 100000;
+        plant->flags.bits.is_shrub = true;
+    }
+
+    // This is correct except for RICE, DATE_PALM, and underground plants
+    // near pool/river/brook. These have both WET and DRY flags.
+    // Should more properly detect if near surface water feature
+    if (!p_raw->flags.is_set(plant_raw_flags::DRY))
+        plant->flags.bits.watery = true;
+
+    plant->material = options.plant_idx;
+    plant->pos = pos;
+    plant->grow_counter = options.age < 0 ? 0 : options.age;
+    plant->update_order = rand() % 10;
+
+    world->plants.all.push_back(plant);
+    switch (plant->flags.whole & 3) // watery, is_shrub
```

Comment:
this is a little too "magic" and could break if the relevant structure changes. maybe just plain old if else if else if else if would be better here